src/build_system.ml

```diff
@@ -250,59 +255,78 @@ module Alias0 = struct
-        acc
-        >>>
-        Build.if_file_exists fn
-          ~then_:(Build.path fn >>^ fun _ -> false)
-          ~else_:(Build.arr (fun x -> x)))
+    Build.lazy_no_targets (lazy (
+      File_tree.Dir.fold dir ~traverse_ignored_dirs:false
+        ~init:(Build.return true)
+        ~f:(fun dir acc ->
+          let path = Path.append ctx_dir (File_tree.Dir.path dir) in
+          let fn = stamp_file (make ~dir:path name) in
+          acc
+          >>>
+          Build.if_file_exists fn
+            ~then_:(Build.path fn >>^ fun _ -> false)
+            ~else_:(Build.arr (fun x -> x)))))
 
   let dep_rec t ~loc ~file_tree =
-    let ctx_dir, src_dir = Path.extract_build_context_dir t.dir |> Option.value_exn in
+    let ctx_dir, src_dir =
+      Path.extract_build_context_dir t.dir |> Option.value_exn
+    in
     match File_tree.find_dir file_tree src_dir with
-    | None -> Build.fail { fail = fun () ->
-      Loc.fail loc "Don't know about directory %s!" (Path.to_string_maybe_quoted src_dir) }
+    | None ->
+      Build.fail { fail = fun () ->
+        Loc.fail loc "Don't know about directory %s!"
+          (Path.to_string_maybe_quoted src_dir) }
     | Some dir ->
       dep_rec_internal ~name:t.name ~dir ~ctx_dir
       >>^ fun is_empty ->
       if is_empty && not (is_standard t.name) then
-        Loc.fail loc "This alias is empty.\n\
-                      Alias %S is not defined in %s or any of its descendants."
+        Loc.fail loc
+          "This alias is empty.\n\
+           Alias %S is not defined in %s or any of its descendants."
           t.name (Path.to_string_maybe_quoted src_dir)
 
   let dep_rec_multi_contexts ~dir:src_dir ~name ~file_tree ~contexts =
-    match File_tree.find_dir file_tree src_dir with
-    | None ->
+    let open Build.O in
+    let dir = find_dir_specified_on_command_line ~dir:src_dir ~file_tree in
+    Build.all (List.map contexts ~f:(fun ctx ->
+      let ctx_dir = Path.(relative build_dir) ctx in
+      dep_rec_internal ~name ~dir ~ctx_dir))
+    >>^ fun is_empty_list ->
+    let is_empty = List.for_all is_empty_list ~f:(fun x -> x) in
+    if is_empty && not (is_standard name) then
       die "From the command line:\n\
-           @{<error>Error@}: Don't know about directory %s!" (Path.to_string_maybe_quoted src_dir)
-    | Some dir ->
-      let open Build.O in
-      Build.all (List.map contexts ~f:(fun ctx ->
-        let ctx_dir = Path.(relative build_dir) ctx in
-        dep_rec_internal ~name ~dir ~ctx_dir))
-      >>^ fun is_empty_list ->
-      let is_empty = List.for_all is_empty_list ~f:(fun x -> x) in
-      if is_empty && not (is_standard name) then
-        die "From the command line:\n\
-             @{<error>Error@}: Alias %s is empty.\n\
-             It is not defined in %s or any of its descendants."
-          name (Path.to_string_maybe_quoted src_dir)
-
-  let default     = make "DEFAULT"
+           @{<error>Error@}: Alias %s is empty.\n\
```

Comment:
To print things consistently, let' use `%S` for the alias name.

Reply:
updated